source/Calamari.AzureResourceGroup/DeployAzureResourceGroupBehaviour.cs

```diff
@@ -32,197 +41,139 @@ public DeployAzureResourceGroupBehaviour(TemplateService templateService, IResou
-            // We increase the poll interval according to the Fibonacci sequence, up to a maximum of 30 seconds.
-            var currentPollWait = 1;
-            var previousPollWait = 0;
-            var continueToPoll = true;
-            const int maxWaitSeconds = 30;
-
-            while (continueToPoll)
+            log.Info("Polling for deployment completion...");
+            try
             {
-                await Task.Delay(TimeSpan.FromSeconds(Math.Min(currentPollWait, maxWaitSeconds)));
-
-                log.Verbose("Polling for status of deployment...");
-                using (var armClient = await createArmClient())
-                {
-                    var deployment = await armClient.Deployments.GetAsync(resourceGroupName, deploymentName);
-                    if (deployment.Properties == null)
-                    {
-                        log.Verbose("Failed to find deployment.Properties");
-                        return;
-                    }
-
-                    log.Verbose($"Provisioning state: {deployment.Properties.ProvisioningState}");
-                    switch (deployment.Properties.ProvisioningState)
-                    {
-                        case "Succeeded":
-                            log.Info($"Deployment {deploymentName} complete.");
-                            log.Info(GetOperationResults(armClient, resourceGroupName, deploymentName));
-                            CaptureOutputs(deployment.Properties.Outputs?.ToString(), variables);
-                            continueToPoll = false;
-                            break;
-
-                        case "Failed":
-                            throw new CommandException($"Azure Resource Group deployment {deploymentName} failed:\n" + GetOperationResults(armClient, resourceGroupName, deploymentName));
-
-                        case "Canceled":
-                            throw new CommandException($"Azure Resource Group deployment {deploymentName} was canceled:\n" + GetOperationResults(armClient, resourceGroupName, deploymentName));
-
-                        default:
-                            if (currentPollWait < maxWaitSeconds)
-                            {
-                                var temp = previousPollWait;
-                                previousPollWait = currentPollWait;
-                                currentPollWait = temp + currentPollWait;
-                            }
-
-                            break;
-                    }
-                }
+                var deploymentResult = await AsyncResourceGroupPollingTimeoutPolicy.ExecuteAsync(async timeoutCancellationToken =>
+                                                                                                 {
+                                                                                                     var delayStrategy = DelayStrategy.CreateExponentialDelayStrategy(TimeSpan.FromSeconds(1), PollingTimeout);
```

Comment:
Is `DelayStrategy.CreateExponentialDelayStrategy` a built-in way for the armClient to retry/poll the state of the operation? If so then you'd want to have a separate timeout for your `AsyncResourceGroupPollingTimeoutPolicy` because I assume you meant that as the absolute timeout. At the moment, they share the 30 second timeout, however it will take much longer than 30 seconds for the exponential delay strategy to actually reach the maximum 30 second polling interval (e.g. with x2 exponential 1+2+4+8+16+30=61). 

I can see that `AzureAppServiceZipDeployBehaviour` has a polling timeout of 3 minutes so perhaps that would be a sensible figure to use.

Reply:
Ah good catch, I was trying to replicate a similar behaviour that was used previously i.e. a fibonacci step off.  

Ill update `AsyncResourceGroupPollingTimeoutPolicy` to 3 minutes which will result in a maximum timeout of 3 minutes and it should make around 10 api calls(Assuming the deployment takes that long and my math is right) which I think is all good.